components/director/internal/destinationfetchersvc/service.go

```diff
@@ -149,38 +158,58 @@ func (d *DestinationService) mapDestinationsToTenant(ctx context.Context, tenant
 	ctx = persistence.SaveToContext(ctx, tx)
 	defer d.Transactioner.RollbackUnlessCommitted(ctx, tx)
 
-	for _, destination := range destinations {
-		correlationID := correlationIDPrefix + destination.CommunicationScenarioID
-		bundles, err := d.BundleRepo.GetBySystemAndCorrelationID(ctx, tenant, destination.XFSystemName, destination.URL, correlationID)
-
-		if len(bundles) == 0 {
-			log.C(ctx).Infof("No bundles found for system '%s', url '%s', correlation id '%s'", destination.XFSystemName, destination.URL, correlationID)
-			continue
-		}
-
-		if err != nil {
-			log.C(ctx).WithError(err).Errorf("Failed to fetch bundle for system '%s', url '%s', correlation id '%s', tenant id '%s'",
-				destination.XFSystemName, destination.URL, correlationID, tenant)
-			continue
-		}
-
-		for _, bundle := range bundles {
-			id := d.UUIDSvc.Generate()
-			revision := d.UUIDSvc.Generate()
-			if err := d.Repo.Upsert(ctx, destination, id, tenant, bundle.ID, revision); err != nil {
-				log.C(ctx).WithError(err).Errorf(
-					"Failed to insert destination with name '%s' for bunlde '%s' and tenant '%s' to DB",
-					destination.Name, bundle.ID, tenant)
-				continue
-			}
-		}
+	if err := dbCalls(); err != nil {
+		log.C(ctx).WithError(err).Error("Failed to execute database calls")
+		return err
 	}
 
 	if err = tx.Commit(); err != nil {
 		log.C(ctx).WithError(err).Error("Failed to commit database transaction")
 		return err
 	}
 	return nil
+
+}
+
+func (d *DestinationService) deleteMissingDestinations(ctx context.Context, revision, tenant string) error {
+	return d.transaction(ctx, func() error {
+		if err := d.Repo.Delete(ctx, revision, tenant); err != nil {
+			log.C(ctx).WithError(err).Errorf("Failed to delete removed destinations for tenant '%s'", tenant)
+		}
+
+		return nil
+	})
+}
+
+func (d *DestinationService) mapDestinationsToTenant(ctx context.Context, tenant, revision string, destinations []model.DestinationInput) error {
+	return d.transaction(ctx, func() error {
+		for _, destination := range destinations {
+			correlationID := correlationIDPrefix + destination.CommunicationScenarioID
+			bundles, err := d.BundleRepo.GetBySystemAndCorrelationID(ctx, tenant, destination.XFSystemName, destination.URL, correlationID)
+
+			if len(bundles) == 0 {
+				log.C(ctx).Infof("No bundles found for system '%s', url '%s', correlation id '%s'", destination.XFSystemName, destination.URL, correlationID)
+				continue
+			}
+
+			if err != nil {
```

Comment:
Shouldn't we check this error first and then `if len(bundles) == 0`?

Reply:
Yes probably a better idea. I think the other way around works as well since bundles will most likely be nil, but let's not risk it.